### PR TITLE
[SPARK-35679][SQL] instantToMicros overflow

### PR DESCRIPTION
**Pull request by dgd-contributor**
### Why are the changes needed?
With Long.minValue cast to an instant, secs will be floored in function microsToInstant and cause overflow when multiply with Micros_per_second 

```
def microsToInstant(micros: Long): Instant = {
  val secs = Math.floorDiv(micros, MICROS_PER_SECOND)
  // Unfolded Math.floorMod(us, MICROS_PER_SECOND) to reuse the result of
  // the above calculation of `secs` via `floorDiv`.
  val mos = micros - secs * MICROS_PER_SECOND  <- it will overflow here
  Instant.ofEpochSecond(secs, mos * NANOS_PER_MICROS)
}
```

But the overflow is acceptable because it won't produce any change to the result

However, when convert the instant back to micro value, it will raise Overflow Error

```
def instantToMicros(instant: Instant): Long = {
  val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND) <- It overflow here
  val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
  result
}
```

Code to reproduce this error
```
instantToMicros(microToInstant(Long.MinValue))
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Test added

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -385,8 +385,10 @@ object DateTimeUtils {
    * microseconds where microsecond 0 is 1970-01-01 00:00:00Z.
    */
   def instantToMicros(instant: Instant): Long = {
-    val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND)
-    val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
+    // we don't need to take care overflow here, since if overflow occurred,
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:388]: I don't think this is the right fix. The range of `instant.getEpochSecond` is  [-31557014167219200L, 31556889864403199L]

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Yes, I've found that too, so I've made a fixed condition for the Long.minValue, could you re-review that for me. Thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -385,9 +385,15 @@ object DateTimeUtils {
    * microseconds where microsecond 0 is 1970-01-01 00:00:00Z.
    */
   def instantToMicros(instant: Instant): Long = {
-    val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND)
-    val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
-    result
+    if (instant.getEpochSecond < 0) {
+      val us = Math.multiplyExact(instant.getEpochSecond + 1, MICROS_PER_SECOND)
+      val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano - NANOS_PER_SECOND))
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]:
```suggestion
      val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano) - MICROS_PER_SECOND)
```

**Comment by gengliangwang** [comment]
As I mentioned in https://github.com/apache/spark/pull/32814#issuecomment-856857466
> We can fix it but it is too corner. The fix makes performance worse, too.


I will leave this one to other committers. cc @MaxGekk @cloud-fan

**Comment by dgd-contributor** [comment]
> As I mentioned in [#32814 (comment)](https://github.com/apache/spark/pull/32814#issuecomment-856857466)
> 
> > We can fix it but it is too corner. The fix makes performance worse, too.
> 
> I will leave this one to other committers. cc @MaxGekk @cloud-fan

Thanks

**Review by cloud-fan — APPROVED** [review]
LGTM. I don't think perf is a big concern in this code path. And the overhead seems very small.

**Comment by MaxGekk** [comment]
@gengliangwang Thank for the ping. I will review this today.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 3]
```diff
@@ -820,4 +820,9 @@ class DateTimeUtilsSuite extends SparkFunSuite with Matchers with SQLHelper {
       }
     }
   }
+
+  test("SPARK-35679: handled overflow error raised by instantToMicros") {
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala:824]: SPARK-35679: instantToMicros should be able to return microseconds of Long.MinValue

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 4]
```diff
@@ -385,9 +385,15 @@ object DateTimeUtils {
    * microseconds where microsecond 0 is 1970-01-01 00:00:00Z.
    */
   def instantToMicros(instant: Instant): Long = {
-    val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND)
-    val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
-    result
+    if (instant.getEpochSecond < 0) {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:388]: @dgd-contributor Could you follow the approach from:
https://github.com/apache/spark/blob/2c8ced95905b8a9f7b98c4913a385d25da455e7a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L872-L885
and handle especially just small range. So, perf impact should be smaller.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala (hunk 3 above)
Comment (dgd-contributor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 4 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Ok done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 5]
```diff
@@ -385,9 +385,16 @@ object DateTimeUtils {
    * microseconds where microsecond 0 is 1970-01-01 00:00:00Z.
    */
   def instantToMicros(instant: Instant): Long = {
-    val us = Math.multiplyExact(instant.getEpochSecond, MICROS_PER_SECOND)
-    val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
-    result
+    val secs = instant.getEpochSecond
+    if (secs == Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)) {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:389]: Could you introduce a constant for `Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 6]
```diff
@@ -385,9 +385,16 @@ object DateTimeUtils {
...
+      val us = Math.multiplyExact(secs + 1, MICROS_PER_SECOND)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: We know `sec` in advance, so, we can calculate `us` in advance too. Please, fold the arithmetic expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 7]
```diff
@@ -385,9 +385,16 @@ object DateTimeUtils {
...
+      val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano) - MICROS_PER_SECOND)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:391]: Are you making some assumption about `instant.getNano`? See the assert the code for intervals:
https://github.com/apache/spark/blob/2c8ced95905b8a9f7b98c4913a385d25da455e7a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L877-L879
Should we have similar assert here?

**Comment by MaxGekk** [comment]
@dgd-contributor Regarding PR's description. Could you fix indentations in code examples according to the code style.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: ok thanks

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: Are you sure? as us might be (secs+1) or secs depend on case

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 7 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:391]: I think it's not nesscessary

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 5 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:389]: Ok, I will

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: `Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)` is constant, right?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: I mostly care of the comparison:
```scala
secs == Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)
```
which is performed per every input. Probably, JVM could optimize it but I would help to it and introduced a constant.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 8]
```diff
@@ -385,9 +385,16 @@ object DateTimeUtils {
...
+      result
+    } else {
+      val us = Math.multiplyExact(secs, MICROS_PER_SECOND)
+      val result = Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))
+      result
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:396]: Let's remove `result`, and just leave `Math.addExact(us, NANOSECONDS.toMicros(instant.getNano))`. The same is for the above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:390]: Yea this should be a constant, but after we enter the if branch and reach here, it's rare and I'm OK to not have a constant for `val us = ...`

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:392]: Done, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 9]
```diff
@@ -378,16 +378,26 @@ object DateTimeUtils {
       throw QueryExecutionErrors.cannotCastUTF8StringToDataTypeError(s, TimestampType)
     }
   }
+  // See issue SPARK-35679
+  // LongMinValueDividedByMicroPerSecond is equal to
+  // Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)
+  // which is -9223372036855L
+  private final val LongMinValueDividedByMicroPerSecond = -9223372036855L
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:385]: `private val MIN_SECONDS = Math.floorDiv(Long.MinValue, MICROS_PER_SECOND)`

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Thanks, I have made change

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@dgd-contributor Does the issue exist in other versions: 3.0, 3.1?

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master, 3.1 and 3.0.
Thank you @dgd-contributor , and @cloud-fan for your review.

**Comment by dgd-contributor** [comment]
Thanks @MaxGekk @cloud-fan @gengliangwang for supporting

**Pull request closed without merging**